lib/IMAP/IMAPClientFactory.php

```diff
@@ -113,10 +116,15 @@ public function getClient(Account $account, bool $useCache = true): Horde_Imap_C
 		);
 		if ($useCache && $this->cacheFactory->isAvailable()) {
 			$params['cache'] = [
-				'backend' => new Cache([
-					'cacheob' => $this->cacheFactory->createDistributed(md5((string)$account->getId())),
-				])];
+				'backend' => $this->hordeCacheFactory->newCache(
+					$account,
+					$this->cacheFactory->createDistributed(md5((string)$account->getId())),
+				),
+			];
 		} else {
+			// WARNING: This is very dangerous! We **will** miss changes when using QRESYNC without
```

Comment:
Means: Don't use mail without redis/apcu?

We added Horde_Imap_Client_Cache_Backend_Null with https://github.com/nextcloud/mail/pull/6880 to mitigate performance/memory issue with larger mailboxes. 

Without redis/apcu larger mailboxes still work, but you have a chance of seeing those ghost messages. 

"Very dangerous" => should we disable the null cache again?

Reply:
Hmm, now I see the problem. Without a real cache we will miss unsolicited changes because Horde will only inform the null cache and don't persist changes anywhere. `QRESYNC` has side effect on all commands because it is used each time an mailbox is opened (`EXAMINE`d).


I think we should implement a "thin" cache that only persist changed messages (so it should only contain the logic I added in this PR to the cache class).

This way we will still prevent the memory leak while still being informed about unsolicited sync changes.

I think the only case where the cache is forcibly disabled in our code is the initial sync which should be fine.



I removed the whole else block in https://github.com/nextcloud/mail/pull/10038 but perhaps this needs second thoughts. Maybe allowing the cache to be fully disabled for the special case of the repair sync which *must* be run without any cache whatsoever to ensure no `QRESYNC` being used.

Reply:
Hmm, now I see the problem. Without a real cache we will miss unsolicited changes because Horde will only inform the null cache and don't persist changes anywhere. `QRESYNC` has side effects because it is used each time a mailbox is opened.

> "Very dangerous" => should we disable the null cache again?

This would be the easiest solution. But perhaps we should implement a "thin" cache that only persists vanished messages (so it should only contain the logic I added in this PR to the cache class). This way we will still prevent said memory leak while still being informed about unsolicited sync changes. And we won't require OCP's cache to be available.

---

PS: I removed the whole else block in https://github.com/nextcloud/mail/pull/10038 but perhaps this needs second thoughts. Maybe allowing the cache to be fully disabled for this one special case of the repair sync. It **must** be run without any cache whatsoever to ensure `QRESYNC` being disabled.

Reply:
As per our discussion, I removed all traces of the old in-memory Horde cache. Only relevant information for QRESYNC is fetched from the db now. That would be the list of all cached uids and the sync token (to extract values of `highestmodseq` and `uidvalidity`).

Otherwise, the new cache class will behave like the built-in null cache in a sense that it won't persist any incoming changes.